woocommerce/woocommerce-ios
# [Mobile Payments] Bring PaymentIntent back and hide implementation details of payment collection

Closes #3927, #3930

⚠️ This PR is against the feature branch implementing support for Card Present Payments, not against develop ⚠️
⚠️ Please note there are several parts of the implementation that are not production ready.⚠️

We continue building our prototype (for reference, p91TBi-4q4-p2), this time, by rolling back some changes we did [in previous PRs](https://github.com/woocommerce/woocommerce-ios/pull/3847/commits/47cd4acd4be6f6ca862c552630ac99746ba7067f) and doing a bit of cleanup in the service API.

This PR does not introduce any user facing changes.


## Changes
* Revert the removal of PaymentIntent, and associated objects and tests. A few weeks back we removed this object, but after looking into what is necessary to render receipts, it is clear that we need it back. (p91TBi-4Ob-p2 for reference)
* Avoid leaking some of the implementation details of the Stripe Terminal SDK, in particular, the fact that processing a payment requires creating an intent, then collecting a payment method for that intent, and finally capturing the payment. Now, all that is done by the service via a new public method that takes the intent parameters as a parameter and returns a publisher that will complete with the final PaymentIntent or an error. In a nutshell, we moved most of the implementation of CardPresentPaymentStore.collectPayment to a method in StripeCardReaderService. 
* Removed a few tests that are just not possible now. #3864 , which goes next, will open up testing again.

## How to test
Testing is a bit complicated because:
1. Requires setting up a store for [Card Present Payment Testing](https://github.com/woocommerce/woocommerce-ios/blob/issue/3748-intent-service/docs/stripe-tests.md#integration-tests) (also P91TBi-4BH-p2 for more specific instructions)  
2. Requires an external hardware reader.

Once the store is setup and there is a hardware reader available:
1. Checkout the branch. Run `bundle exec pod install`. 
2. Build and run on a device. 
3. Switch an external card reader on. The only card reader supported by now is the BBPOS Chipper 2X BT.
4. Log in to an account that matches a store that has been configured for testing.
5. Navigate to Settings in the app. (Tap the gear button in the top right)
6. In the "Store Settings" section, tap "Manage Card Reader".
7. Make sure the card reader is on and the blue light is blinking (we don't do any error handling yet, so it's important that things are setup for the happy path)
8. Tap "Connect card reader"
9. Wait until the app suggests one reader to connect to.
10. Tap "Connect to reader"
11. Make sure the reader light turns solid blue.
12. Navigate to an order (any order with payment method Cash On Delivery).
13. Scroll down to the "Collect Payment" button
14. Tap the button. Follow the instructions on screen

The result should look like this:

https://user-images.githubusercontent.com/2722505/114028174-2de44780-9846-11eb-9155-bab847453f3e.mov


Update release notes:
